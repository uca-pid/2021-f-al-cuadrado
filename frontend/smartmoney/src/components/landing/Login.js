import React from 'react';
import {useState} from 'react';
import "./style.css";
import webStyles from "./webStyles";
import mobilStyles from "./mobilStyles";
import { useMediaQuery } from 'react-responsive';
import RequiredField from '../RequiredField/requiredField';

const Login = ({setLayoutForgotPassword,setLayoutRegister}) => {
    
  const isMobileDevice = useMediaQuery({
    query: "(max-device-width: 480px)",
  });
    const [user, setUser] = useState('');
    const [password, setPassword] = useState('');
    const [invalidCredentials, setInvalidCredentials] = useState('none');
    const [userEmpty, setUserEmpty] = useState(false);
    const [passwordEmpty, setPasswordEmpty] = useState(false);
   
    const login = () => {
      setInvalidCredentials('none')
      if(user==='')setUserEmpty(true);
      if(password==='')setPasswordEmpty(true);
      if(!(user===''||password==='')){
        const requestOptions = {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ email: user, password: password})
        };
        fetch('https://smart-money-back.herokuapp.com/login/', requestOptions)
          .then(response => response.json())
          .then(data => {
            localStorage.setItem('session',JSON.stringify(data));
            window.location.href = "./home"
          })
          .catch(err => setInvalidCredentials('block'))
      }
    }

    const isEmpty = (input, isEmpty) => {
        if(input==='')isEmpty(true)
      }


    return(
        <div className="formContainer" name="Login">     
            <form className="form">
              <p className="invalidCredentials" style={{display:invalidCredentials}}>Credenciales incorrectas</p>
              <p className="label" >Usuario</p>
              <input style={isMobileDevice ? (userEmpty ? mobilStyles.inputEmpty : mobilStyles.input) : (userEmpty ? webStyles.inputEmpty : webStyles.input)} type="text" value={user} onChange={e => setUser(e.target.value)} onFocus={()=>setUserEmpty(false)} onBlur={()=>isEmpty(user,setUserEmpty)}/>
              {userEmpty&&<RequiredField/>}
              <p className="label" >Contraseña</p>
              <input style={isMobileDevice ? (passwordEmpty ? mobilStyles.inputEmpty : mobilStyles.input) : (passwordEmpty ? webStyles.inputEmpty : webStyles.input)} type="password" value={password} onChange={e => setPassword(e.target.value)} onFocus={()=>setPasswordEmpty(false)} onBlur={()=>isEmpty(password,setPasswordEmpty)}/>
              {passwordEmpty&&<RequiredField/>}
              <input style={isMobileDevice ? mobilStyles.forgotPassword : webStyles.forgotPassword} type="button" onClick={setLayoutForgotPassword} value="Olvidé mi contraseña" />
              <input 
                className = "button1"
                type="button" 
                onClick={login} 
                value="Iniciar sesión" 
                disabled={userEmpty||passwordEmpty}/>
              <div className="line"></div>
              <input 
                className = "button2"
                type="button" 
                onClick={setLayoutRegister} 
                value="Registrarse" />
            </form>
          </div> 
    )
}

export default Login;