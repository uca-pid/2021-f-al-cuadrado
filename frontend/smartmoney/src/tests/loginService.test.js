import React from 'react'
import Login from '../../src/components/landing/Login'
import {render, screen, fireEvent} from '@testing-library/react';

jest.mock('../../src/services/loginService');

describe('Login', () => {

  const mockFuntion = () => {return true};
  let wrapper;
  beforeEach(() => {
    wrapper = render(<Login setLayoutForgotPassword={mockFuntion} setLayoutRegister={mockFuntion}/>)
  })

  it("login component renders sucsessfully", () => {
    expect(wrapper).toMatchSnapshot();
  })

  it("login set inputs", () => {
    fireEvent.change(screen.getByTestId("user-input"),{target:{value:"test@gmail.com"}})
    fireEvent.change(screen.getByTestId("password-input"),{target:{value:"12345"}})
    expect(screen.getByTestId("user-input").value).toBe("test@gmail.com")
    expect(screen.getByTestId("password-input").value).toBe("12345")
  })

  it("login successful", () => {
    fireEvent.change(screen.getByTestId("user-input"),{target:{value:"test@gmail.com"}})
    fireEvent.change(screen.getByTestId("password-input"),{target:{value:"12345"}})
    
    fireEvent.click(screen.getByTestId("login-button"));
    expect(localStorage.getItem('session')).toBe(JSON.stringify({code:"123456", user_id:1}))
  })

  it("login wrong email", () => {
    fireEvent.change(screen.getByTestId("user-input"),{target:{value:"wrongEmail@gmail.com"}})
    fireEvent.change(screen.getByTestId("password-input"),{target:{value:"12345"}})
    
    expect(screen.getByText("Wrong credentials").style._values.display).toBe("none");
    fireEvent.click(screen.getByTestId("login-button"));
    expect(screen.getByText("Wrong credentials").style._values.display).toBe("block");
  })

  it("login wrong password", () => {
    fireEvent.change(screen.getByTestId("user-input"),{target:{value:"test@gmail.com"}})
    fireEvent.change(screen.getByTestId("password-input"),{target:{value:"wrongPassword"}})
    
    expect(screen.getByText("Wrong credentials").style._values.display).toBe("none");
    fireEvent.click(screen.getByTestId("login-button"));
    expect(screen.getByText("Wrong credentials").style._values.display).toBe("block");
  })


})
