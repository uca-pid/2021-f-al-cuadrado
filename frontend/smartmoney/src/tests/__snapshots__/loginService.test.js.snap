// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login login component renders sucsessfully 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="formContainer"
        name="Login"
      >
        <form
          class="form"
        >
          <p
            class="invalidCredentials"
            style="display: none;"
          >
            Wrong credentials
          </p>
          <div
            class="MuiFormControl-root MuiFormControl-marginDense MuiTextField-root css-4mzek5-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
            >
              User
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="user-input"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-dhh0px"
                >
                  <span>
                    User
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </p>
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-marginDense MuiTextField-root css-4mzek5-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
            >
              Password
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="password-input"
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-dhh0px"
                >
                  <span>
                    Password
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </p>
          </div>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root css-1knaqv7-MuiButtonBase-root-MuiButton-root"
            style="justify-content: left; font-size: 10px;"
            tabindex="0"
            type="button"
          >
            Forgot password
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="css-1m4faer-MuiStack-root"
            style="min-width: 70%;"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="login-button"
              style="min-width: 50%;"
              tabindex="0"
              type="button"
            >
              Login
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
              style="max-width: 700px;"
              tabindex="0"
              type="button"
            >
              Register
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="formContainer"
      name="Login"
    >
      <form
        class="form"
      >
        <p
          class="invalidCredentials"
          style="display: none;"
        >
          Wrong credentials
        </p>
        <div
          class="MuiFormControl-root MuiFormControl-marginDense MuiTextField-root css-4mzek5-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
          >
            User
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="user-input"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-dhh0px"
              >
                <span>
                  User
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginDense MuiTextField-root css-4mzek5-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
          >
            Password
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="password-input"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-dhh0px"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </p>
        </div>
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root css-1knaqv7-MuiButtonBase-root-MuiButton-root"
          style="justify-content: left; font-size: 10px;"
          tabindex="0"
          type="button"
        >
          Forgot password
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="css-1m4faer-MuiStack-root"
          style="min-width: 70%;"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-sghohy-MuiButtonBase-root-MuiButton-root"
            data-testid="login-button"
            style="min-width: 50%;"
            tabindex="0"
            type="button"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            style="max-width: 700px;"
            tabindex="0"
            type="button"
          >
            Register
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
